Controller UI support

PiersSK opened the pull request:
### Controller Support

The game should now be fully playable with a controller, and will instantly switch between controller/MK upon any input. This ended up being a BIG change.

This primarily effects all the UI navigation
- Diary Terminal
- Bridge Terminal
- Server Room Terminal
- Pause Menu
- Main Menu

### Key things to test:
- When you're on any UI, try putting an input from the other source (i.e. if you're using controller, nudge your mouse).
    - If you're moving TO mouse and keyboard, the cursor should be visible, button prompts should disappear and no particular buttons/inputs should be highlighted.
    - If moving TO controller then the cursor should disappear, button prompts should appear and SOMETHING should be selected. **IF NOTHING IS SELECTED YOU BECOME STUCK SO THIS IS VERY IMPORTANT**. If a UI has subscreens, please test on all screens

- Main Menu
    - Can you start the game correctly??
- Pause Menu
    - Can you navigate between all 5 options correctly (main nav buttons)? They should loop
    - Resume game should exit the pause menu, as should pressing ButtonEast (B or Circle)
    - Audio/Input settings should open the settings panel
    - When on a settings panel, you should be able to use left and right to change the sliders
    - When on a settings panel, you shouldn't be able to scroll the main nav buttons, but pressing ButtonEast should close the settings panel and return you to the main nav buttons
    - Exit game should open a confirm choice screen and you should start highlighting Cancel. ButtonEast should also change to back out to the pause menu
- Diary Terminal
    - Rubbish with controller whilst free inputs are in but you should be able to press ButtonSouth (A or Cross) to highlight it white and theoretically that'd open an on screen keyboard if the device supports it
- Bridge Terminal
    - You should be able to toggle between the two buttons - very simple!
- Server Terminal
    - You should be able to navigate between the buttons easily. If you've inserted the **CM Override PD** into the correct server it adds another button. Please do this and check navigation still works
    - Opening any subscreen with a button should highlight a button to continue navigation (same for backing out)
    - Admin screen should highlight text box with a green square to suggest you're ready to insert, pressing ButtonSouth would open on screen keyboard if possible

PiersSK committed:
feat(code): highlights text inputs and improves diary

PiersSK committed:
feat(code): adds UI navigation to bridge terminal

Also moves deselect duty to general UI Manager

PiersSK committed:
feat(code): adds controller navigation to server terminal

PiersSK committed:
fix(code): changes some server navigation routes

PiersSK committed:
feat(code): adds PD screen navigation

PiersSK committed:
fix(code): bridge terminal navigation set explicitly

PiersSK committed:
feat(code): clearer highlighting on selected buttons

PiersSK committed:
feat(code): adds fully controller supported pause menu

PiersSK committed:
fix(code): buttons detatch listeners on disable

PiersSK committed:
fix(code): create method for checking if gamepad

PiersSK committed:
feat(code): improves UI transitions with MK + Gamepad

PiersSK committed:
feat(code): improves diary terminal input navigation

PiersSK committed:
feat(code): standardises input highlighting

PiersSK committed:
feat(code): adds back navigation on server terminal

also adds touchscreen keyboard

OrcaOfTime commented:
A couple notes from testing behaviour:

Quiz terminal / typing boxes - Not expecting typing to be available however the triangle button was the 'Enter' equivalent. Whether or not that is intended behaviour it wasn't indicated clearly in the UI at the bottom left then same way other input instruciton were i.e. 'R1 to look out window'

Main terminal - no issues, clear indication of what is being selected on screen, select button works 

Pause menu - Will appreciate a second opinion on this but the navigation of pause menu felt a bit off. The subsections feel like too much navigation to get to the settings rather than just being able to navigate through all six options (as of writing). 
Would it be significantly difficult to have the selected bar highlighted with a box and then select to able to edit and confirm change so player can select the individual settings?
Also the x button not being visible for the 'Return to game' and 'Return to main menu' but visible for the sub section selection feels inconsistent. The highlighting is already looks clear so a generic 'X to select' feels like it would suffice.

Main menu - No navigation available for main menu to select 'Play'

PiersSK commented:
> A couple notes from testing behaviour:
> 
> Quiz terminal / typing boxes - Not expecting typing to be available however the triangle button was the 'Enter' equivalent. Whether or not that is intended behaviour it wasn't indicated clearly in the UI at the bottom left then same way other input instruciton were i.e. 'R1 to look out window'
> 
> Main terminal - no issues, clear indication of what is being selected on screen, select button works
> 
> Pause menu - Will appreciate a second opinion on this but the navigation of pause menu felt a bit off. The subsections feel like too much navigation to get to the settings rather than just being able to navigate through all six options (as of writing). Would it be significantly difficult to have the selected bar highlighted with a box and then select to able to edit and confirm change so player can select the individual settings? Also the x button not being visible for the 'Return to game' and 'Return to main menu' but visible for the sub section selection feels inconsistent. The highlighting is already looks clear so a generic 'X to select' feels like it would suffice.
> 
> Main menu - No navigation available for main menu to select 'Play'

I have no idea why triangle is submit - I can't find a way to change it 😂 

Very fair feedback on the pause menu, it was definitely designed to navigate a menu laid out for M+K. I'll have a think about some ways to rearrange it to make it more controller friendly!

PiersSK committed:
feat(code): reworks pause menu for better navigation

PiersSK committed:
fix(code): fixes incorrect bindings

- buttonNorth on submit not wanted
- re-adds analogue stick to navigation

PiersSK committed:
feat(code): refactors inputManager to seperate from player

also removes any remaining local references and refers to it as a singleton in all cases

PiersSK committed:
feat(code): main menu controller support

PiersSK commented:
> A couple notes from testing behaviour:
> 
> Quiz terminal / typing boxes - Not expecting typing to be available however the triangle button was the 'Enter' equivalent. Whether or not that is intended behaviour it wasn't indicated clearly in the UI at the bottom left then same way other input instruciton were i.e. 'R1 to look out window'
> 
> Main terminal - no issues, clear indication of what is being selected on screen, select button works
> 
> Pause menu - Will appreciate a second opinion on this but the navigation of pause menu felt a bit off. The subsections feel like too much navigation to get to the settings rather than just being able to navigate through all six options (as of writing). Would it be significantly difficult to have the selected bar highlighted with a box and then select to able to edit and confirm change so player can select the individual settings? Also the x button not being visible for the 'Return to game' and 'Return to main menu' but visible for the sub section selection feels inconsistent. The highlighting is already looks clear so a generic 'X to select' feels like it would suffice.
> 
> Main menu - No navigation available for main menu to select 'Play'

All of these things should be fixes! Have removed a submit button on text boxes for now. Can re-add if on screen keyboard testing later down the line does not autosubmit. (descoped from this as we may remove them)

OrcaOfTime commented:
Layout looks a lot better, makes it much easier to navigate, can confirm main menu buttons are selectable. Some bugs I noticed in testing.
Controller used - PS5 Dualsense
Steps - 
  1. Press pause button to enter pause menu 
  2. Select a settings sub-menu i.e. sound. This will change the 'select' button prompt on bottom left of screen to the 'change setting value' button prompt
  3. Exit pause menu using pause button instead of using circle to back out of sub-menu and then back out again or selecting 'return to game'

Results - 
1. The button prompt to change a setting's value remains on screen, even updating it's UI when changing from console remote to mouse and keyboard
2. Opening pause menu again shows the sub-menu open though it is not selecting, I would expect either the sub-menu to be closed or the last highlighted setting upon unpausing to be highlighted again.

Player must back out of pause menu using the 'return' button prompt (circle in this case) to clear the button prompt UI properly.

OrcaOfTime committed:
Merge branch 'develop' into controller-ui-support

OrcaOfTime commented:
Have made a separate bug ticket for described behaviour above

JamesYJ95 commented:
Hey hi howdy! Couple things - I tested with both my PS5 DualSense and my Switch Pro Controller (unsupported I know ehe)

**DualSense**
1. I was able to reproduce Sean's issue re: pause menu
2. I spotted that on the Select PD to insert menu, the cursor appears
3. Plays fuckin' slick otherwise

**Pro Controller**
(Discussed on call but writing here for a record) - my PC treats my Pro Controller as both a PC Input AND a GamePad. As a result, some inputs go a little crazy (click left stick causes on-screen keyboard to appear), moving right stick causes wild sensitivity because it's being used as both a mouse and an analog stick.

PiersSK committed:
fix(code): correctly resets pause state on exit

PiersSK committed:
fix(code): corrects PD menu input switch logic
